### PR TITLE
Manipulate column row

### PR DESCRIPTION
**Pull request by davidfig**
- [x] insert column
- [x] insert row
- [x] delete column
- [x] delete row
- [x] fix formula updates
- [x] validations
- [x] delete multiple column and rows
- [x] fix right-click on grid headings
- [x] fix delete row not working
- [x] copy formatting from origin column/row
- [x] insert/delete into sheet offsets

### CODE REVIEW

File: quadratic-core/src/controller/active_transactions/pending_transaction.rs [hunk 1]
```diff
@@ -90,10 +108,16 @@ impl Default for PendingTransaction {
             complete: false,
             generate_thumbnail: false,
             cursor_undo_redo: None,
-            send_validations: HashSet::new(),
+            validations: HashSet::new(),
```
Comment (ddimaria) [inline quadratic-core/src/controller/active_transactions/pending_transaction.rs:111]: Better name 🙌

File: quadratic-core/src/controller/execution/control_transaction.rs [hunk 2]
```diff
@@ -114,11 +115,74 @@ impl GridController {
...
+                                x: pos.x,
+                                y: pos.y,
+                                html: None,
+                                w: None,
+                                h: None,
+                            });
+                            if let Ok(html) = serde_json::to_string(&html) {
+                                crate::wasm_bindings::js::jsUpdateHtml(html);
+                            } else {
+                                dbgjs!("Error serializing html");
```
Comment (ddimaria) [inline quadratic-core/src/controller/execution/control_transaction.rs:161]: Add the html and/or pos with this error message?

File: quadratic-core/src/controller/execution/run_code/mod.rs [hunk 3]
```diff
@@ -114,66 +77,14 @@ impl GridController {
...
-                            sheet_id.to_string(),
-                            sheet_pos.x,
-                            sheet_pos.y,
-                            Some(code_cell),
-                            Some(render_code_cell),
-                        );
-                    }
-                }
-            }
+            transaction.add_from_code_run(sheet_id, pos, &old_code_run);
```
Comment (ddimaria) [inline quadratic-core/src/controller/execution/run_code/mod.rs:81]: How are you differentiating operations from reverse operations here?

Comment (ddimaria) [reply quadratic-core/src/controller/execution/run_code/mod.rs:81]: Ignore, I see below.

File: quadratic-core/src/grid/column.rs [hunk 4]
```diff
@@ -363,6 +375,101 @@ impl<B: BlockContent> ColumnData<B> {
...
+        let mut new_blocks = BTreeMap::new();
+
+        for (start, block) in self.0.iter() {
+            // block is before the insertion point, then copy
+            if *start < y && (*start + block.len() as i64) < y {
+                new_blocks.insert(*start, block.clone());
+            }
+            // block is at or after the insertion point, then shift right
+            else if *start >= y {
+                let mut new_block = block.clone();
```
Comment (ddimaria) [inline quadratic-core/src/grid/column.rs:393]: Not important, but I use `clone()` when writing the first pass at code, then convert to `to_owned()` to denote to my self (and others) that I went through and ensured I need a copy (you obviously need a copy here).

File: quadratic-core/src/grid/column.rs [hunk 5]
```diff
@@ -363,6 +375,101 @@ impl<B: BlockContent> ColumnData<B> {
...
+        self.0 = new_blocks;
+        changed
+    }
+
+    /// Removes a position and shifts the remaining positions to the left.
+    pub fn remove_and_shift_left(&mut self, y: i64) -> bool {
+        let mut changed = false;
+        let mut new_blocks = BTreeMap::new();
+
+        for (start, block) in self.0.iter() {
```
Comment (ddimaria) [inline quadratic-core/src/grid/column.rs:421]: If you change to `into_iter()` you may be able to remove the clone below.

File: quadratic-core/src/grid/sheet/borders/borders_col_row.rs [hunk 6]
```diff
@@ -0,0 +1,914 @@
+//! Inserts and removes columns and rows for borders. Also provides fn to get
+//! undo operations for these changes.
+
+use itertools::Itertools;
+
+use crate::{controller::operations::operation::Operation, grid::SheetId, selection::Selection};
+
+use super::{BorderStyleCellUpdates, Borders};
+
+impl Borders {
```
Comment (ddimaria) [inline quadratic-core/src/grid/sheet/borders/borders_col_row.rs:10]: Good use of iterators in this file ❤️

**Comment by ddimaria** [comment]
I finished reviewing the Rust files (all looks good), but I'm holding off on the JS review as there is still work being done there.

File: quadratic-client/src/app/gridGL/UI/gridHeadings/GridHeadings.ts [hunk 7]
```diff
@@ -13,6 +13,8 @@ import { getColumnA1Notation } from './getA1Notation';
 
 type Selected = 'all' | number[] | undefined;
 
+export type IntersectsHeadings = { column?: number; row?: number; corner?: true };
```
Comment (ddimaria) [inline quadratic-client/src/app/gridGL/UI/gridHeadings/GridHeadings.ts:16]: ❤️

File: quadratic-core/src/grid/column.rs (hunk 4 above)
Comment (davidfig) [reply quadratic-core/src/grid/column.rs:393]: Sounds like a good method. I'm probably too lazy for it, though.

File: quadratic-core/src/grid/column.rs (hunk 5 above)
Comment (davidfig) [reply quadratic-core/src/grid/column.rs:421]: doesn't that clone the entire old record?

File: quadratic-core/src/grid/sheet/bounds.rs [hunk 8]
```diff
@@ -221,6 +221,28 @@ impl Sheet {
         }
     }
 
+    /// Returns the lower and upper bounds of formatting in a row, or `None` if
+    /// the row has no formatting.
+    pub fn row_bounds_formats(&self, row: i64) -> Option<(i64, i64)> {
+        let column_has_row = |(_x, column): &(&i64, &Column)| column.has_format_in_row(row);
+        let min = if let Some((index, _)) = self.columns.iter().find(column_has_row) {
```
[the remark this answers is not in the document]
Comment (davidfig) [reply quadratic-core/src/grid/sheet/bounds.rs:228]: i'm surprised clippy didn't find this

**Pull request merged**